### ISSUE TITLE
We shouldn't require track transferability

### ISSUE THREAD

**Issue by guidou** [w3c/mediacapture-transform#113]
The current version of the API requires track transferability, but this shouldn't be necessary.
Currently, tracks are useless on workers except for this API, so we shouldn't add that as a requirement.

A way to keep the API worker first which has several benefits is to follow the postMessage-like approach of webrtc-encoded-transform.

Something (subject to discussion) like:

For MediaStreamTrackProcessor:
```
// main
navigator.mediaDevices.createTrackProcessor(myWorker, mytrack, myOptions, [myOptions]);`

// worker
ontrackprocessor = event => {
    let processor = event.processor;
   // process things using 
}
```

For VideoTrackGenerator:
```
// main
let generatedTrack = navigator.mediaDevices.createVideoTrackGenerator(myWorker, myOptions, [myOptions]);`

// worker
ontvideorackgenerator = event => {
    let generator = event.generator;
   // generate frames for `generatedTrack` 
}
```

**Comment by guidou** [comment]
cc @jan-ivar @youennf

**Comment by jan-ivar** [comment]
What problem is this solving? Transferring the track has other benefits like being able to apply constraints and read track stats and settings.

**Comment by guidou** [comment]
It solves the problem that you don't need track transferability to implement this API, which we consider a blocker.
Also, the benefits of transferring a track to a worker are very limited. In fact, I would argue that the only benefit would be using this API. 
You can call applyConstraints and read stats/settings on main, where the main application is.

This is a pattern that we are already using in encoded transform and should easily allow us to have interoperable implementations.

**Comment by youennf** [comment]
One use case for transferring media stream track is to create a track (via VideoTrackGenerator) and send it to sinks like RTCRtpSender or MediaRecorder.

**Comment by guidou** [comment]
The proposal is that createVideoTrackGenerator() is called from Window and returns a promise with a MediaStreamTrack on Window, where the RTCRtpSender or MediaRecorder are. The generator (which no longer has a track field) is created in the worker (the application gets it via an event, just like an RTCRtpScriptTransformer).

This removes the need to transfer the track. You only needed to transfer the track from the worker to window because the current spec creates the track on the worker, where it is largely useless, as all the track APIs are on Window.

**Comment by guidou** [comment]
Another benefit of this API surface is that it allows feature detection on main without creating a worker.

**Comment by jan-ivar** [comment]
This can be feature detected on main [like this](https://jsfiddle.net/jib1/92r6h0qn/):
```js
function isMstTransferable() {
  try {
    const [track] = document.createElement('canvas').captureStream().getVideoTracks();
    new MessageChannel().port1.postMessage(track, [track]);
    return true;
  } catch (e) {
    if (e.name != "DataCloneError") throw e;
    return false;
  }
}
```

**Comment by jan-ivar** [comment]
> It solves the problem that you don't need track transferability to implement this API, which we consider a blocker, at least for the medium term.

@guidou why is transfer a blocker? What do you mean by medium term? Safari has already shipped this and it [works](https://blog.mozilla.org/webrtc/unbundling-mediastreamtrackprocessor-and-videotrackgenerator).

If you explain the problem, perhaps their engineers can help?

**Comment by guidou** [comment]
It's a blocker for Chromium to ship it in the short term since Chromium doesn't implement track transferability and will not have it for quite some time.

I don't expect Chromium to have track transferability in the short term, so I guess we won't have an interoperable API for a long time.

**Comment by guidou** [comment]
> This can be feature detected on main [like this](https://jsfiddle.net/jib1/92r6h0qn/):
> 
> ```js
> function isMstTransferable() {
>   try {
>     const [track] = document.createElement('canvas').captureStream().getVideoTracks();
>     new MessageChannel().port1.postMessage(track, [track]);
>     return true;
>   } catch (e) {
>     if (e.name != "DataCloneError") throw e;
>     return false;
>   }
> }
> ```

This feature-detects track transferability, not mediacapture-transform.
With the current API  you need to create a worker to feature-detect, which is costly and unergonomic.

**Comment by jan-ivar** [comment]
I doubt that's needed. As you said, _"tracks are useless on workers except for this API"_. If MST transfer is detected, it seems reasonable to assume some purpose awaits these tracks in the worker.

This works [in the only current implementation](https://webkit.org/blog/15443/news-from-wwdc24-webkit-in-safari-18-beta/#webrtc): _"WebKit for Safari 18 beta adds support for MediaStreamTrack processing in a dedicated worker."_

This seems like a property worth emulating. I've added [a note](https://bugzilla.mozilla.org/show_bug.cgi?id=1749543#c2) to Firefox's implementation bug to do the same. Thanks for bringing attention to this!

**Comment by jan-ivar** [comment]
> ... Chromium doesn't implement track transferability and will not have it for quite some time.

If there's some difficulty or problem with the spec's [transfer steps](https://w3c.github.io/mediacapture-extensions/#transferable-mediastreamtrack) as specified, please bring it to our attention so we can address it.

**Comment by jan-ivar** [comment]
> You can call applyConstraints and read stats/settings on Window, where the main application is.

Yes, but waiting on postMessage for these measurements hardly seems ideal. In the current spec, the worker transform can inspect real-time [track stats](https://w3c.github.io/mediacapture-extensions/#dom-mediastreamtrackvideostats) counters like `deliveredFrames`, `discardedFrames` and `totalFrames` synchronously, and correlate them with the `VideoFrame` it is currently processing.

**Comment by guidou** [comment]
> > You can call applyConstraints and read stats/settings on Window, where the main application is.
> 
> Yes, but waiting on postMessage for these measurements hardly seems ideal. 

That goes the other way too if you want access to the track on Window (which is the more common case today).

> In the current spec, the worker transform can inspect real-time [track stats](https://w3c.github.io/mediacapture-extensions/#dom-mediastreamtrackvideostats) counters like `deliveredFrames`, `discardedFrames` and `totalFrames` synchronously, and correlate them with the `VideoFrame` it is currently processing.

I'm not opposed to supporting transferability.  I'm opposed to making it a requirement to use mediacapture-transform, as that will have the practical consequence of delaying interoperable implementations.

We already have a pattern for adding worker support without requiring transferability of tracks or streams. This doesn't mean applications are forbidden from transferring tracks on browsers that support it if they want to.
It just means that applications that don't need to transfer tracks to do processing (which are most if not all applications today) can more quickly have an interoperable API in practice.

**Comment by jan-ivar** [comment]
> That goes the other way too if you want access to the track on Window ...

No, because tracks can be cloned. With transfer, stats are readily available in both places. So the problem of a transformer needing a roundtrip to main to read settings and applyConstraints, for lack of transfer, would be new with this proposal.

**Comment by jan-ivar** [comment]
> I'm not opposed to supporting transferability.

Great! Since you said tracks are useless on workers except for the worker API,  does this mean you support the worker API?

> I'm opposed to making it a requirement to use mediacapture-transform, ...

It already is a requirement.

> ... as that will have the practical consequence of delaying interoperable implementations.

I doubt attempting to standardize a third new API and waiting for three implementations will get us to interop quicker.

Safari has shipped, and Firefox is working on it. 1½ < 3 + one WG. I've filed https://github.com/w3c/mediacapture-extensions/issues/158 to help.

Creating a permanent web API to solve one implementer's short-term scheduling seems against [§ 1.7. Add new capabilities with care](https://w3ctag.github.io/design-principles/#new-features) and [§ 1.9. Leave the web better than you found it](https://w3ctag.github.io/design-principles/#leave-the-web-better).

> This doesn't mean applications are forbidden from transferring tracks on browsers that support it if they want to.

Having web developers navigate between 3 instead of 2 different APIs to do the same thing sounds worse, not better.

**Comment by youennf** [comment]
> I said tracks are useless on workers except for this API (i.e., mediacapture-transform) which artificially requires them.

This is not artificial, transferring a track to a worker has real benefits compared to the approach you mention.
Let's take the example of a web application wanting to do background blur on a camera feed via a MediaStreamTrackProcessor and a VideoTrackGenerator.

First, lifetime management is easier.
When the VideoTrackGenerator track gets stopped, its WritableStream will be closed. The web application can listen to this via its closed promise and call `stop` on the getUserMedia track.
Also, stopping the worker will kill both VideoTrackGenerator and getUserMedia track, housekeeping is simpler :)
This is less convenient when the WritableStream lives in a different context than the track, web developer will need to post message.

Second, configuration management.
If the getUserMedia track is muted, the web app will likely want to mute the VideoTrackGenerator.
Ditto when getUserMedia track is unmuted.
If the getUserMedia track is in the same context as VideoTrackGenerator, it is very easy to implement for the web developer.
Otherwise, web app has to postMessage.

This has a real user consequences: a few frames will likely be missed by VideoTrackGenerator when getUserMedia track gets unmuted if the web app has to postMessage. With the worker approach, missing frames would be a bug in the UA implementation.

The same principle applies to `configurationchange`, `getSettings`, `applyConstraints`.
It is much easier for VideoTrackGenerator, MediaStreamTrackGenerator and getUserMedia track to be all in the same context to make use of these APIs.

Finally, we introduced MediaStreamTrack transferability as a way to cover some longer term use cases (grabbing camera in an iframe but do rendering/processing in another iframe). The current spec is more future-proof from that point of view as well.

> User needs come before the needs of web page authors, which come before the needs of user agent implementors, which come before the needs of specification writers, which come before theoretical purity.

Right, I think user needs will likely be better served with the current API, as described above.
I tend to agree that track transferability requires more work from UA implementors, but these costs are overweighted by user and web developer benefits.

> I also doubt an API that ignores developer requirements and concerns by at least one implementor will get us to interop.

What are the developer requirements that have been ignored?
So far, the developer feedback we received is that MSTP and VTG are working fine in Safari.

**Comment by guidou** [comment]
> > I said tracks are useless on workers except for this API (i.e., mediacapture-transform) which artificially requires them.
> 
> This is not artificial, transferring a track to a worker has real benefits compared to the approach you mention. Let's take the example of a web application wanting to do background blur on a camera feed via a MediaStreamTrackProcessor and a VideoTrackGenerator.
> 

Yes, it is an artificial *requirement*.

If you have a use case where having the track in the worker is useful, then that can be very valid, but it doesn't justify making transferability it a *requirement* for mediacapture-transform.  I didn't say track transferability is an artificial feature. I'm just saying it is an artificial *requirement* for  mediacapture-transform that, in addition, is often detrimental.

> First, lifetime management is easier. When the VideoTrackGenerator track gets stopped, its WritableStream will be closed. The web application can listen to this via its closed promise and call `stop` on the getUserMedia track. Also, stopping the worker will kill both VideoTrackGenerator and getUserMedia track, housekeeping is simpler :) This is less convenient when the WritableStream lives in a different context than the track, web developer will need to post message.
> 

You don't need transferability *as a requirement* to support this use case. 
UAs that support track transferability can perfectly support this use case even if transferability is not a requirement for mediacapture-transform.
BTW, I have never heard of this use case from actual developers, but I am not opposed to it. 

> Second, configuration management. If the getUserMedia track is muted, the web app will likely want to mute the VideoTrackGenerator. Ditto when getUserMedia track is unmuted. If the getUserMedia track is in the same context as VideoTrackGenerator, it is very easy to implement for the web developer. Otherwise, web app has to postMessage.
> 
> This has a real user consequences: a few frames will likely be missed by VideoTrackGenerator when getUserMedia track gets unmuted if the web app has to postMessage. With the worker approach, missing frames would be a bug in the UA implementation.
> 

I don't think this is an actual problem because if the getUserMedia track is muted, it will produce no frames and the VideoTrackGenerator will see no frames.
Still, if you want to support this use case in this manner, there is nothing in this proposal preventing it.
Like I said, you don't need transferability as a requirement to support this. You just need transferability, which no one is opposing.


> The same principle applies to `configurationchange`, `getSettings`, `applyConstraints`. It is much easier for VideoTrackGenerator, MediaStreamTrackGenerator and getUserMedia track to be all in the same context to make use of these APIs.
> 

The same applies. I haven't heard developers request this, but even if it's a useful use case, you don't need transferability as a requirement to support this. You just need transferability.

> Finally, we introduced MediaStreamTrack transferability as a way to cover some longer term use cases (grabbing camera in an iframe but do rendering/processing in another iframe). The current spec is more future-proof from that point of view as well.
> 

That is an actual use case for which I have seen developer demand and IMO is the main value track transferability can provide. This is completely independent of having track transferability as a requirement for mediacapture-transform.

> > User needs come before the needs of web page authors, which come before the needs of user agent implementors, which come before the needs of specification writers, which come before theoretical purity.
> 
> Right, I think user needs will likely be better served with the current API, as described above.

User needs are not better served by having transferability as a requirement.
Eliminating the requirement of transferability does not prevent any of the use cases mentioned before.
For that you just need transferability, not transferability as a requirement for mediacapture-transform.

On the other hand, *requiring* transferability does make things difficult for some common use cases. The most obvious is playing both the gUM and VTG tracks on a video element in a before/after effects view. In this case you no longer have the gUM track on Window and therefore can't play it on an element. The same applies if you want to use any other track sink available only on Window.

>  I tend to agree that track transferability requires more work from UA implementors, but these costs are overweighted by user and web developer benefits.
> 

All the arguments I've seen so far are benefits of transferability as a standalone feature.  None of these benefits are derived from that transferability being a *requirement* for mediacapture-transform.
Moreover, I've presented use cases where transferability as a requirement makes it more difficult to support common use cases.

So, transferability as a standalone feature supports both the use cases you presented and the ones I presented, but transferability as a mediacapture-transform requirement only supports the use cases you presented and fails to properly support the ones I presented. 
It is clear to me that removing the transferability as a requirement for mediacapture-transform better serves the needs of developers.


> > I also doubt an API that ignores developer requirements and concerns by at least one implementor will get us to interop.
> 
> What are the developer requirements that have been ignored? So far, the developer feedback we received is that MSTP and VTG are working fine in Safari.

Here are some developer requirements that are well known to us and which are ignored by the current version of the spec (not all of these are related to the issue we're discussing which is track transferability as a mediacapture-transform requirement):
* Audio support
* Keeping the gUM track on Window while processing on Worker (before/after view and other use cases that require track sinks available only on Window)
* Easy feature detection (without requiring the creation of a Worker)
* Processing on Window.

**Comment by youennf** [comment]
> Here are some developer requirements that are well known to us and which are ignored by the current version of the spec (not all of these are related to the issue we're discussing which is track transferability as a mediacapture-transform requirement):

Let's only talk about the requirements that are relevant to this particular issue (audio support and processing on window are out of scope).

> * Keeping the gUM track on Window while processing on Worker (before/after view and other use cases that require track sinks available only on Window)

The before/after view can be implemented by transferring a clone of the track instead of transferring the track itself.

> * Easy feature detection (without requiring the creation of a Worker)

@jan-ivar provided a feature detection approach that works in Safari (and will likely work in Firefox).
This is good enough in practice, feature detection does not have to be technically pure.

I am sympathetic to the needs of browser implementors. So far though, I haven't seen new information that warrants revisiting the design of this API.
Also, this API shape has mostly remained untouched for several years, probably since the spec first public working draft and reached consensus at the time within the WebRTC WG. This API shipped in one UA, and is being implemented in another UA.

**Comment by jan-ivar** [comment]
> The specific change proposed in this issue is not about "short-term" scheduling. It is to make the API better.
If a use case can be solved appropriately without introducing a dependency on another feature, then it is better to solve it without introducing that dependency. The fact that it results in an API that is easier to implement is a consequence of that design being better.

I think this API is shortsighted. It's tightly coupled, artificially tied to main thread, and it reinvents postMessage.

Our goal is to enable MediaStreamTrack processing in dedicated workers. This might include MediaStreamTracks originating in the worker someday, e.g. from an [OffscreenCanvas.captureStream()](https://github.com/w3c/mediacapture-fromelement/issues/65#issuecomment-1956065210) or other sources already exposed in the worker. Or an RTCDataChannel in a worker feeding a VideoTrackGenerator created there.

Since [we all agree](https://github.com/w3c/mediacapture-transform/issues/113#issuecomment-2413866184) MediaStreamTracks will exist in workers eventually, the simplest API is the one that accepts them there.

The idiomatic way to get data to workers is with postMessage, using [transferable objects](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects) if needed.

So I disagree we shouldn't depend on other web platform features. It's doing it all ourselves that's the mistake. At least that's how I read [§ 1.7 Add new capabilities with care](https://www.w3.org/TR/design-principles/#new-features).

**Comment by guidou** [comment]
I thought the plan was to summarize our positions in a separate issue and ask TAG for their opinion, but here's my reply.

> > The specific change proposed in this issue is not about "short-term" scheduling. It is to make the API better.
> > If a use case can be solved appropriately without introducing a dependency on another feature, then it is better to solve it without introducing that dependency. The fact that it results in an API that is easier to implement is a consequence of that design being better.
> 
> I think this API is shortsighted. It's tightly coupled, artificially tied to main thread, 
>
This API is not tighly coupled with anything. If a developer wants to transfer a track to a Worker and manage all its state and lifetime there, there is nothing in the proposed API preventing it. Just like nothing prevents developers from managing the track on Window if that is what they prefer. 

The one that forces developers to use track transferability even if they'd rather not use it is the one that tightly couples two features that should be independent of each other.

>
> and it reinvents postMessage.
> 
This API does not reinvent postMessage anymore than encoded transform does. If this is such a bad thing, should I file an issue in encoded transform to eliminate the same pattern there and require that RTCRtpSender and RTCRtpReceiver (or some other object) be transferrable too?

> Our goal is to enable MediaStreamTrack processing in dedicated workers. This might include MediaStreamTracks originating in the worker someday, e.g. from an [OffscreenCanvas.captureStream()](https://github.com/w3c/mediacapture-fromelement/issues/65#issuecomment-1956065210) or other sources already exposed in the worker. Or an RTCDataChannel in a worker feeding a VideoTrackGenerator created there.

All this can be supported without tightly coupling mediacapture-transform with track transferability.
I don't think it is possible to find a single use case where forcing the user to use track transferability is better than allowing it, but without forcing it.

> 
> Since [we all agree](https://github.com/w3c/mediacapture-transform/issues/113#issuecomment-2413866184) MediaStreamTracks will exist in workers eventually, the simplest API is the one that accepts them there.
>

No it's not the simplest API. It is a lot more complex to tightly couple two features that should be independent.
It is not only more complex for UA implementors, which cannot develop the features independently; but, more importantly, it is more complex for Web developers, who are forced to use an unnecessary feature and complex workarounds to solve otherwise nonexisting problems.
Now Web developers are forced to clone a track, transfer one of them, and introduce track management logic in two separate realms even if their preference would be to do all track management on Window.

Even more importantly, the proposed API does not even need to be a replacement for the existing one.
Since this API removes the tight coupling between both features, it isn't really much of a problem to provide the constructors in the existing API as a convenience for hypothetical applications that would prefer to do all track management in the worker.
 
> The idiomatic way to get data to workers is with postMessage, using [transferable objects](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects) if needed.
> 

Again I ask, why is this a problem. Is encoded transform non-idiomatic? Should we eliminate the RTCRtpScriptTransform constructor and introduce a new transferable object there to be used with postMessage, or make senders and receivers transferable? 
Or is it a problem here, but not there?

> So I disagree we shouldn't depend on other web platform features. It's doing it all ourselves that's the mistake.
The mistake is to force a dependency on another feature that should be independent.  
I'd like to see a single use case where this dependency provides a benefit for web developers compared to having the features be orthogonal.

> At least that's how I read [§ 1.7 Add new capabilities with care](https://www.w3.org/TR/design-principles/#new-features).
We read it very differently. Adding dependencies between features that should be orthogonal and forcing developers to use complex workarounds to deal with those unnecessary dependencies is,, in my view, the opposite of adding capabilities with care.

**Comment by jan-ivar** [comment]
@guidou I appreciate your efforts to simplify the API, but I believe your proposal introduces more complexity rather than reducing it. It seems unclear whether your proposed API is intended to replace the existing MediaCapture Transform API or to coexist alongside it.

If it's meant to coexist, then we're asking developers to navigate between multiple APIs that achieve similar goals, which can lead to confusion and fragmentation. This also increases the burden on browser implementers to support multiple APIs, delaying interoperability.

If it's meant to replace the existing API, it disregards the implementations already shipped in Safari and in progress in Firefox, which would fragment the ecosystem further and negate the developer feedback we've already received.

Moreover, your proposal doesn't seem to stand on its own because it doesn't cover all the use cases the current API does — particularly future scenarios where tracks originate in workers or need to be fully managed within a worker context.

Requiring track transferability isn't an unnecessary dependency; it's a design choice that provides significant benefits to developers, such as simplified lifetime and configuration management, as well as access to track stats and settings directly within the worker.

Adding another API also goes against the web platform design principles of keeping the platform consistent and avoiding unnecessary complexity.

I believe it's better for us to focus on implementing the existing API consistently across browsers and addressing any implementation challenges together, rather than introducing an alternative that could fragment the ecosystem.

**Comment by guidou** [comment]
> @guidou I appreciate your efforts to simplify the API, but I believe your proposal introduces more complexity rather than reducing it. 

Can you elaborate on how is it more complex? Especially for Web developers.
One way would be to compare how intended use cases are implemented with each API. 

> It seems unclear whether your proposed API is intended to replace the existing MediaCapture Transform API or to coexist alongside it.

It can be both. I would prefer replace.

> If it's meant to coexist, then we're asking developers to navigate between multiple APIs that achieve similar goals, which can lead to confusion and fragmentation. 

It would be better to replace, but since there is one implementation, coexist seems acceptable.

> This also increases the burden on browser implementers to support multiple APIs, delaying interoperability.
It's not uncommon for an API to have multiple constructors or factory methods to serve different use cases.  
In this case, I'm proposing factory methods that the provide the following benefits:
* It properly supports all use cases that have been presented so far. In particular it is better at supporting the common use case of an application wanting to manage the track Window and media processing on Worker.
* It removes the tight coupling between two separate API. This has the advantage that they can be developed independently by UA implementers.

> If it's meant to replace the existing API, it disregards the implementations already shipped in Safari and in progress in Firefox, which would fragment the ecosystem further and negate the developer feedback we've already received.

For this reason, coexist would be acceptable.

> Moreover, your proposal doesn't seem to stand on its own because it doesn't cover all the use cases the current API does — particularly future scenarios where tracks originate in workers or need to be fully managed within a worker context.

I'm talking about real use cases deployed in production right now, not hypothetical ones that might never exist. I believe the former should have more weight than the latter in the design of the WG's APIs.

> Requiring track transferability isn't an unnecessary dependency;

Requiring track transferability for use cases that don't need it is indeed an unnecessary dependency.

> it's a design choice that provides significant benefits to developers, such as simplified lifetime and configuration management, as well as access to track stats and settings directly within the worker.

There is no simplified lifetime and configuration management. Any use case that prefers to manage track lifetime on Window (basically all use cases deployed today) requires much more complex lifetime management with the existing API.

> Adding another API also goes against the web platform design principles of keeping the platform consistent and avoiding unnecessary complexity.

Requiring track transferability for use cases that don't need it is precisely adding unnecessary complexity.

> I believe it's better for us to focus on implementing the existing API consistently across browsers and addressing any implementation challenges together, rather than introducing an alternative that could fragment the ecosystem.

The ecosystem is already fragmented. This proposal might have the side effect of making it easier to reduce that fragmentation as it makes it possible for UA implementors to develop two features independently using patterns that are already implemented and tested. Forcing two features than can (and should) be orthogonal to have a dependency such that one has to be implemented before the other does nothing to help reduce the already existing fragmentation.

Finally, I think we have reached the point in which we are just repeating the same arguments without achieving consensus. Shouldn't we go ahead with the plan to file our positions in separate issues and get TAG's input?

**Comment by jan-ivar** [comment]
> Again I ask, why is this a problem. Is encoded transform non-idiomatic?

Encoded transform is bespoke.

> Should we eliminate the RTCRtpScriptTransform constructor and introduce a new transferable object there to be used with postMessage, or make senders and receivers transferable?

No, because unique tradeoffs were involved, and that [FPWD](https://lists.w3.org/Archives/Public/public-webrtc/2021Sep/0000.html) has already shipped in two browsers.

> Or is it a problem here, but not there?

I believe it's on the person filing the issue to produce a convincing problem that needs fixing. Otherwise I see no new information since [FPWD](https://lists.w3.org/Archives/Public/public-webrtc/2022Feb/0024.html) that warrants revisiting the design of this API.

Usage of the spec API seems fine, as seen in [this blog](https://blog.mozilla.org/webrtc/unbundling-mediastreamtrackprocessor-and-videotrackgenerator).

**Comment by guidou** [comment]
> > Again I ask, why is this a problem. Is encoded transform non-idiomatic?
> 
> Encoded transform is bespoke.
> 

In what way that doesn't apply here?


> > Should we eliminate the RTCRtpScriptTransform constructor and introduce a new transferable object there to be used with postMessage, or make senders and receivers transferable?
> 
> No, because unique tradeoffs were involved, and that [FPWD](https://lists.w3.org/Archives/Public/public-webrtc/2021Sep/0000.html) has already shipped in two browsers.
> 

There is a similar tradeoff here. There is nothing magical about FPWDs such that they cannot be improved.


> > Or is it a problem here, but not there?
> 
> I believe it's on the person filing the issue to produce a convincing problem that needs fixing. Otherwise I see no new information since [FPWD](https://lists.w3.org/Archives/Public/public-webrtc/2022Feb/0024.html) that warrants revisiting the design of this API.
> 

The use case where the application needs the track on Window is a convincing one.
Basically, that describes all applications using MediaStreamTrack on the Web today.


> Usage of the spec API seems fine, as seen in [this blog](https://blog.mozilla.org/webrtc/unbundling-mediastreamtrackprocessor-and-videotrackgenerator).

Doesn't seem particularly fine to me. 
Properly supporting lots of real-world applications using MediaStreamTrack on Window would seem better to me.

**Comment by guidou** [comment]
> > There is nothing magical about FPWDs such that they cannot be improved.
> 
> Well, we are very far from FPWD, the spec reached consensus within the WebRTC WG and has been very stable for a few years now. This design cannot come as surprise since both editors of this spec are from Google. Also, there will be soon two implementations. This should prove that the current spec is implementable and precise enough to get interop. This would allow to move the spec to REC.
> 

That there are two implementations doesn't prevent changing the spec to improve it. We do that all the time with specs that are much more mature than this one.

> > The use case where the application needs the track on Window is a convincing one.
> 
> So far, I have not seen a usecase where the API you propose is providing more than the existing API. Could you be more specific?
> 

All existing applications  today (including the ones that do video processing) currently do so with the tracks on Window, where they manage all the logic. Forcing them to migrate to manage the track in the worker is an unnecessary barrier.


> For instance, the API you are proposing is most probably shimable using the current API. The reverse is not true.
> 
> Taking a real example to compare the two APIs as an exercise, we could take the use case of doing real time encoding and sending of a video track to the network (using data channel, web transport, or the future encoded source proposal). This requires the web page to potentially adapt the frame rate and/or resolution to the network conditions.
>
> With the current API, the adaptation logic is all happening solely in the worker via `applyConstraints`. With the proposed API, the web application would have to postMessage to the window environment, which is doable but more complex and suboptimal.
> 

That sounds more like a hypothetical example than a real one. Either way, my proposal supports it fine. There is nothing that prevents transferring the track to the worker in my proposal, so saying that some use case works better when you transfer the track is not an argument against the proposal.

You would have to show evidence that forcing the transfer of the track is better in all, or at least most use cases, than making it optional.


> > Doesn't seem particularly fine to me.
> > Properly supporting lots of real-world applications using MediaStreamTrack on Window would seem better to me.
> 
> Could you precise what use case and what advantages you see? My understanding is that both APIs will roughly have a similar level of functionality.

The spec version forces a pattern of transferring a track and managing its lifetime there, while historically all applications manage tracks on Window. My proposal makes it easy to support both. Forces the transfer of the track to the worker is a not necessarily a good pattern, especially for existing applications.

The use case is all existing applications.

**Comment by youennf** [comment]
> That there are two implementations doesn't prevent changing the spec to improve it

What is proposed is an entire rewriting the API/WebIDL, which means rewriting a large part of the spec.

> The use case for managing the track on Window is all existing applications.

It would help immensely if we have a solid proof that the new API is hard to use.
So far, it does not seem to be the case.

AIUI, the current API and the proposed API have the same feature set, so there should be no impact on end users. All we are debating is ease of use of the two APIs.

We should compare this potential ease-of-use benefit with the known amount of work building the new proposal would require:
1. WebRTC WG discussions to reaching consensus on the new API design
2. Rewriting of the spec
3. Deprecation and removal of the current API (we surely do not want to have two APIs for the same job)
4. Reimplementations in all browsers.

This is a big ask.

**Comment by guidou** [comment]
> > That there are two implementations doesn't prevent changing the spec to improve it
> 
> What is proposed is an entire rewriting the API/WebIDL, which means rewriting a large part of the spec.
> 
Nothing of the sort. The proposal is basically adding a new factory method  for MSTP /VTG. No need to remove anything


> > The use case for managing the track on Window is all existing applications.
> 
> It would help immensely if we have a solid proof that the new API is hard to use. So far, it does not seem to be the case.
> 

Safari implements it. Are you aware of any major applications moving from canvas capture to the new API? Can you share any adoption numbers?
Maybe we can also conduct some sort of developer survey.

> AIUI, the current API and the proposed API have the same feature set, so there should be no impact on end users. All we are debating is ease of use of the two APIs.
> 
The proposed API is just an additional factory method so that you xan keep the tracks on Window, which will make it easier to migrate existing applications and develop new ones following commonly used patterns withvMediaStreamTracks

> We should compare this potential ease-of-use benefit with the known amount of work building the new proposal would require:
> 
> 1. WebRTC WG discussions to reaching consensus on the new API design
> 2. Rewriting of the spec
> 3. Deprecation and removal of the current API (we surely do not want to have two APIs for the same job)
> 4. Reimplementations in all browsers.
> 
> This is a big ask.

The needs of users and Web developers are above the needs of browser implementers and spec writers in the priority of constituencies so I don't see this as a blocker.
Also, in terms of implementation, the proposal reuses a pattern already implemented in browsers for encoded transform.

**Comment by youennf** [comment]
> Nothing of the sort. The proposal is basically adding a new factory method for MSTP /VTG. No need to remove anything

If we are not removing the existing API, why adding an API that can be shimed on the existing API?

> Safari implements it. Are you aware of any major applications moving from canvas capture to the new API?

This is still early as Safari shipped it recently.
There is active interest and, so far I haven't heard of any blocker or difficulty for using the API.

**Comment by guidou** [comment]
> > Nothing of the sort. The proposal is basically adding a new factory method for MSTP /VTG. No need to remove anything
> 
> If we are not removing the existing API, why adding an API that can be shimed on the existing API?
> 

It removes the transferability requirement and makes it easy to do feature detection without creating a worker.


> > Safari implements it. Are you aware of any major applications moving from canvas capture to the new API?
> 
> This is still early as Safari shipped it recently. There is active interest and, so far I haven't heard of any blocker or difficulty for using the API.

It would be good to ask those potential users if they would prefer an API that forces to transfer tracks to a worker or one that lets them choose.between Worker and Window.

**Comment by jan-ivar** [comment]
> > Nothing of the sort. The proposal is basically adding a new factory method for MSTP /VTG. No need to remove anything
> 
> If we are not removing the existing API, why adding an API that can be shimed on the existing API?

It's a bit more than a factory method. For comparison, here's a trivial factory function ([works](https://jsfiddle.net/jib1/94Lmqse0/) in Safari):
```js
async function createVideoTrackGeneratorAndProcessor(worker, track) {
  const before = track.clone();
  worker.postMessage({before}, [before]);
  const {data} = await new Promise(r => worker.addEventListener("message", r));
  return data.after;
}
```
But this differs from what @guidou seems to want: the worker reacting to `track.applyConstraints()` on the main thread (or  wherever it is transferred I suppose?) That doesn't exist today.

Having processing react to tracks on other threads seems more complex than requiring the track to be on the same thread (Safari [stops](https://jsfiddle.net/jib1/cuyo3h25/) processing if the `before` track is transferred away).

It feels like we haven't thought past main thread here. Asking websites to postMessage constraints seems simpler.

**Comment by mehagar** [comment]
I'm in the process of trying to use Safari's new spec-compliant implementation of MediaStreamTrackProcessor/VideoTrackGenerator right now for a virtual background feature. Our existing app ([https://app.zoom.us/wc](https://app.zoom.us/wc)) handles both Chrome's implementation (we transfer the `readable` to the worker) and the case where MediaStreamTrackProcessor isn't available at all. Here are my thoughts as a user of this new API change:

-  It is indeed annoying to have to feature detect by first creating a Worker. I'm not aware of any other API that is Worker only, and our existing code assumes that all feature detection can be done on the main thread. We would either have to create and then destroy a dummy worker just for feature detection, or significantly modify how our startup flow works in order to feature detect when the real worker is first created.

- The WebRTC/getUserMedia API is for the most part main thread only (with the recent exception of transferrable RTCDataChannels). We had previously made the assumption that calling `applyConstraints` on the main thread would always work, but that is no longer the case with this API change. I guess the correct way to solve this would be to postMessage constraints to the worker, but this would require a larger change as we would need to have different code paths for browsers that support transferrable tracks and those that don't.

Overall, the API change forces us to restructure how we manage tracks by changing how track state is synced between the main thread and worker threads. Previously, all track management is done on the main thread - but with this new API it's now all done in the worker threads. For new applications this wouldn't be a problem, but it is for us, where we make many assumptions about how tracks are managed. I understand that there are benefits for this API be available on workers, but not why it needs to be worker *only*, especially when doing so makes adapting existing applications challenging.

**Comment by jan-ivar** [comment]
> I'm in the process of trying to use Safari's new spec-compliant implementation of MediaStreamTrackProcessor/VideoTrackGenerator right now for a virtual background feature.

Glad to hear you're working on this. You have my sympathy for having to deal with these API differences right now. But this goes away as more browsers implement the standard API. Once all major engines have it, absence of this feature should be the rare exception and not worth optimizing for.

I'd like to help you over the hump here, as that seems more productive than trying to standardize a third API over what amounts to a transient problem.

> - It is indeed annoying to have to feature detect by first creating a Worker. I'm not aware of any other API that is Worker only, ...

[AudioWorkletProcessor](https://developer.mozilla.org/en-US/docs/Web/API/AudioWorkletProcessor) is an interface only exposed in worklets. But developers never feature-detect it directly. They pragmatically assume it exists if [AudioWorkletNode](https://developer.mozilla.org/en-US/docs/Web/API/AudioWorkletNode) exists on main-thread.

Similarly, developers today can test for [MST transferability](https://github.com/w3c/mediacapture-transform/issues/113#issuecomment-2383916093) and assume the worker APIs exist if true. This works reliably in all browsers today, and I think it would be foolish for a future browser to implement MST transfer ahead of implementing the two APIs it's for. Firefox has already committed to not doing so.

Optimizing for a feature NOT existing seems of diminishing value over time.

> - ... We had previously made the assumption that calling applyConstraints on the main thread would always work, but that is no longer the case with this API change. I guess the correct way to solve this would be to postMessage constraints to the worker, ...

Yes, did you try fiddle I posted above to _"[postMessage constraints](https://jsfiddle.net/jib1/hwn4L8kj/)"_ in Safari? The salient part:

```js
function apply(height) {
  worker.postMessage({constraints: {height}});
}
range.oninput = () => apply(number.value = range.value);
number.oninput = () => apply(range.value = number.value);

function work() {
  let track;
  onmessage = async ({data: {before, constraints}}) => {
    if (before) {
      // ...
      track = before;
    } else if (constraints && track) {
      await track.applyConstraints(constraints);
    }
  };
```

> ... this would require a larger change as we would need to have different code paths for browsers that support transferrable tracks and those that don't. ... For new applications this wouldn't be a problem, but it is for us, where we make many assumptions about how tracks are managed.

Thank you for acknowledging that this is not a problem for new applications. I don't think this meets the bar for standardizing a new API. That said, I'd like to help.

[Here](https://jsfiddle.net/jib1/zocw86m3/) is a simple shim that modifies the track's applyConstraints function to do the postMessage. It's a bit ugly perhaps, but this seems required for any drop-in to satisfy a set of (otherwise incompatible) previous assumptions. Hope this helps.

**Comment by jan-ivar** [comment]
What flexibility was removed? My [shim](https://jsfiddle.net/jib1/zocw86m3/) suggests the API proposed here is a subset of the existing spec, not a superset.

What major problems? The evidence presented here suggests a minor inconvenience at worst. The spec follows established patterns for workers, which is postMessage, not bespoke cross thread APIs.

**Comment by guidou** [comment]
> What flexibility was removed? 

The flexibility to manage the control surface and the media surface independently.

> My [shim](https://jsfiddle.net/jib1/zocw86m3/) suggests the API proposed here is a subset of the existing spec, not a superset.

The shim hasn't established that. It shows that you can postMessage around and have similar semantics for applyConstraints, which is an asynchronous operation. But that's not necessarily true for all control operations.  For example, `stop()` and `enabled` are synchronous.  Blocking the main thread to provide the same semantics while messaging across threads can have a nontrivial cost in terms of user-perceived performance. Event handlers probably also have their own set of issues.

> What major problems?  The evidence presented here suggests a minor inconvenience at worst. 

Having to redesign applications because they have to abandon the well-established assumption that they can control tracks on Window is not a minor inconvenience. It's evidence that the API is failing to properly support its intended use cases.

> The spec follows established patterns for workers, which is postMessage, not bespoke cross thread APIs.

What you call "bespoke" is a pattern established on the Web by this Working Group for the WebRTC Encoded Transform API.  And among established patterns on the Web, it is not a bad one IMO.

**Comment by youennf** [comment]
> For example, `stop()` and `enabled` are synchronous. Blocking the main thread to provide the same semantics while messaging across threads can have a nontrivial cost in terms of user-perceived performance. Event handlers and `mute` probably also have their own set of issues.

Can you elaborate what the actual issues are? I can understand the general idea but not the practical implications.

In particular, the source is off the main thread and all control surfaces basically need to hop to the source thread.
At worse, the shim would add a slight delay by hopping to the worker thread and then to the source thread.
I don't think the main thread should be blocked to emulate `stop()` or `enabled`.

In the case of a track -> track transform, I would expect applications to `stop()` the transformed track, the worker gets notified in VideoTrackGenerator, which will stop the worker track. No issue AFAIUI.
Ditto for `enabled`, I would expect it to happen on the transformed track (and maybe on the source track as an optimization?).
I could see an extension to VideoTrackGenerator so that it gets notified that all the tracks related to it have `enabled = false` like native sources have as a way to optimise processing (stop sending video frames that no sink is interested in).

**Comment by guidou** [comment]
> > For example, `stop()` and `enabled` are synchronous. Blocking the main thread to provide the same semantics while messaging across threads can have a nontrivial cost in terms of user-perceived performance. Event handlers and `mute` probably also have their own set of issues.
> 
> Can you elaborate what the actual issues are? I can understand the general idea but not the practical implications.
> 
> In particular, the source is off the main thread and all control surfaces basically need to hop to the source thread. At worse, the shim would add a slight delay by hopping to the worker thread and then to the source thread. I don't think the main thread should be blocked to emulate `stop()` or `enabled`.
> 

If you don't block the main thread for sync operations, then you don't get the same semantics as a track on the main thread.  This is to answer the claim that an API that allows the track on the main thread is a subset of one that forces the track to be transferred to the worker because it can be shimmed.
My claim is that a shim that provides the full functionality and semantics of a track on Window is more complex and has more performance issues than what an initial shim that supports async operations like applyConstraints would suggest.

Alternatively,  applications can just abandon the previously valid assumption that all tracks can be controlled from Window and use alternative, more asynchronous and more complex architectures. 

IMO, this is placing a heavy burden on Web developers for something that can be easily fixed by adding a couple of additional factory methods to the API.

**Comment by jan-ivar** [comment]
I think we need to distinguish theoretical differences from ones that matter. My bar for the shim was to minimize refactoring, not design a zero-work drop-in.

All examples presented so far (here and in #116) stop both the original track and the processed track on main thread at the same time, so shimming `stop()` hardly seems necessary.

I agree with @youennf people are going to stop the processed track (which stops the worker clone per https://github.com/w3c/mediacapture-transform/issues/117#issuecomment-2502034790). Same with enabled.

If they don't then yeah the processed track can outlive the original just fine. It's a feature.

**Comment by guidou** [comment]
> I think we need to distinguish theoretical differences from ones that matter. My bar for the shim was to minimize refactoring, not design a zero-work drop-in.
> 
> All examples presented so far (here and in #116) stop both the original track and the processed track on main thread at the same time, so shimming `stop()` hardly seems necessary.

One of the claims here is that the proposed change is not necessary because it is a "subset" that can be shimmed.
It is good that we have now established that this claim is false.

The rest of the arguments don't address the fact that many applications prefer to have the track on Window, since that is how tracks have always worked and provides for a better application architecture in many (probably most) cases.
The argument that it is possible to write the application when the track is in the worker simply does not address this issue.

**Comment by guidou** [comment]
BTW, the proposal of reusing the WebRTC Encoded Transform pattern here is just a possible way to solve the core issue, which is that we should support keeping the track on Window if the application prefers that.
If we solve this using any other mechanisms/API surfaces, that will probably be OK.

**Comment by jan-ivar** [comment]
You don't need to maintain duplicate versions of this state.

`applyConstraints` only ever fails with `OverconstrainedError` on `exact` constraints, which amounts to a camera discovery API. It will faithfully produce the same results regardless of whether it is called in a worker or main-thread.

**Comment by youennf** [comment]
> The API currently in the spec (with a lot of "no consensus" remarks that make it impossible to move that API to PR) is a proposal.

It is more than a proposal. The API reached consensus within the WebRTC WG 3 years ago and the API has received very few API updates during those 3 years. Since then, the API has been validated by implementation and shipped.

> the API currently in the spec may not be the best for developers.

The bar is pretty high to do a full redesign, given what I mentioned above.

The summary  of the discussion AIUI is:
- There is no loss of functionality when using the standard API.
- There may be performance issues with the standard API, it is still unclear where and how much.
- There is some amount of work to do the transition from the proprietaty API to the standard API. The main issue seems to lie around `applyConstraints`. `stop` and `enabled` have been mentioned by @guidou as well. It is not clear what this amount of work is and whether it is unreasonable. @jan-ivar showed a small but partial shim.
- This amount of code may deter existing apps from using that API (these apps may stick to `rvfc` and friends).

**Comment by youennf** [comment]
> * It is not clear what this amount of work is and whether it is unreasonable

It seems three scenarios could be used to evaluate this issue:
- WebCodecs decoding in a worker -> MediaStreamTrack via `VideoTrackGenerator`.
- Background replacement like processing via `MediaStreamTrackProcessor/VideoTrackGenerator`.
- MediaStreamTrack -> WebCodecs encoding in a worker via `MediaStreamTrackProcessor`.

AIUI, scenarios 2 and 3 are probably the most interesting ones.

**Comment by alvestrand** [comment]
> > The API currently in the spec (with a lot of "no consensus" remarks that make it impossible to move that API to PR) is a proposal.
> 
> It is more than a proposal. The API reached consensus within the WebRTC WG 3 years ago and the API has received very few API updates during those 3 years. Since then, the API has been validated by implementation and shipped.

We may have different interpretations of "the API".
In particular, we have not resolved #23 and #29, and we still have the "no consensus" note on main-track exposure in https://w3c.github.io/mediacapture-transform/#track-processor-interface and https://w3c.github.io/mediacapture-transform/#video-generator-interface. The 3 years of relative silence here is an acknowledgement of deadlock, not a consensus.

**Comment by youennf** [comment]
I am referring to "the API" in the scope of this issue: doing out of main thread processing of MediaStreamTrack video frames.
https://github.com/w3c/mediacapture-transform/issues/23 and https://github.com/w3c/mediacapture-transform/issues/29 are orthogonal to this issue.

For that use case, the WG consensus (use track transfer with the worker API) was validated 3 years ago and was not questioned until this issue was filed.

**Comment by dontcallmedom** [comment]
I have a growing feeling that an important part of the issue discussed here is a strong disconnect between track cloning being initially designed to allow and facilitate disconnected track states (being able to apply different constraints to the various clones) and track cloning as the underlying mechanism for transferring tracks making it *hard* to keep synchronized states across transferred clones.

Since track state is fundamentally an off-js state, I wonder if having an explicit API to keep clone states synchronized would alleviate sufficiently many of these particular concerns.

**Comment by jan-ivar** [comment]
Claims of user-perceived performance issues need to be substantiated. As do claims that applyConstraints, stop and enabled are hard transition problems.

**Comment by alvestrand** [comment]
> I have a growing feeling that an important part of the issue discussed here is a strong disconnect between track cloning being initially designed to allow and facilitate disconnected track states (being able to apply different constraints to the various clones) and track cloning as the underlying mechanism for transferring tracks making it _hard_ to keep synchronized states across transferred clones.

Track cloning in order to obtain individually-constrainable tracks is far older than track transfer. Cloning-for-constraints steps being invoked as part of transfer may not be the right way to specify transfer.

**Comment by jan-ivar** [comment]
Again, transfer does not require cloning. Cloning was only offered as a solution to legacy apps wanting access to an unprocessed track on main thread.
